controllers/azurestackedgemachine_controller.go

```diff
@@ -239,117 +239,117 @@ func (r *AzureMachineReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, ret
         networkInterface.VnetName = clusterScope.Vnet().Name
         networkInterface.SubnetName = clusterScope.ControlPlaneSubnet().Name
         networkInterface.VnetResourceGroup = clusterScope.Vnet().ResourceGroup
         networkInterface.StaticIPAddress = clusterScope.APIServerIP()
         networkInterface.IsPrimary = true
-        machineScope.AzureMachine.Status.NetworkInterfaces = append(machineScope.AzureMachine.Status.NetworkInterfaces, networkInterface)
-        machineScope.Info(fmt.Sprintf("Network Interface for mgmt control plane %v", machineScope.AzureMachine.Status.NetworkInterfaces))
+        machineScope.AzureStackEdgeMachine.Status.NetworkInterfaces = append(machineScope.AzureStackEdgeMachine.Status.NetworkInterfaces, networkInterface)
+        machineScope.Info(fmt.Sprintf("Network Interface for mgmt control plane %v", machineScope.AzureStackEdgeMachine.Status.NetworkInterfaces))
         //update the status on the machine
         if err := machineScope.Close(ctx); err != nil {
             machineScope.Info("Error in updating machine status")
-            r.Recorder.Eventf(machineScope.AzureMachine, corev1.EventTypeWarning, "Error in updating AzureMachine status", "Failed to update AzureMachine status")
+            r.Recorder.Eventf(machineScope.AzureStackEdgeMachine, corev1.EventTypeWarning, "Error in updating AzureStackEdgeMachine status", "Failed to update AzureStackEdgeMachine status")
             return reconcile.Result{}, err
         }
-        machineScope.Info(fmt.Sprintf("Successfully updated AzureMachine status"))
+        machineScope.Info(fmt.Sprintf("Successfully updated AzureStackEdgeMachine status"))
 
     }
 
-    machineScope.Info(fmt.Sprintf("ReconciledIP value is %t", machineScope.AzureMachine.Status.ReconciledIP))
-    if machineScope.AzureMachine.Status.ReconciledIP == false {
-        if err:= r.reconcileAzureMachineIPAddress(machineScope, ctx, clusterScope, azureIPPoolScope); err!= nil {
-            r.Recorder.Eventf(machineScope.AzureMachine, corev1.EventTypeWarning, "Error in reconcileAzureMachineIPAddress", errors.Wrapf(err, "error reconciling IP for machine %s", machineScope.Name()).Error())
-            return reconcile.Result{}, errors.Wrapf(err, "Error in reconcileAzureMachineIPAddress")
+    machineScope.Info(fmt.Sprintf("ReconciledIP value is %t", machineScope.AzureStackEdgeMachine.Status.ReconciledIP))
+    if machineScope.AzureStackEdgeMachine.Status.ReconciledIP == false {
+        if err:= r.reconcileAzureStackEdgeMachineIPAddress(machineScope, ctx, clusterScope, azureStackEdgeIPPoolScope); err!= nil {
+            r.Recorder.Eventf(machineScope.AzureStackEdgeMachine, corev1.EventTypeWarning, "Error in reconcileAzureStackEdgeMachineIPAddress", errors.Wrapf(err, "error reconciling IP for machine %s", machineScope.Name()).Error())
+            return reconcile.Result{}, errors.Wrapf(err, "Error in reconcileAzureStackEdgeMachineIPAddress")
         }
     }
     
```

Comment:
If the IP is assigned to the VM, reconcileAzureStackEdgeMachineIPAddress and later on service crashes before it got a chance to run newAzureStackEdgeMachineService. Will we end up leaking IP address

Reply:
When we call reconcileAzureStackEdgeMachineIPAddress, we do the following:
1) Get free IP from the free IP Pool
2) Add to allocated IP Pool
3) Build network interface object
4) Modify the azureippool resource 
5) set the ReconciledIP to true on the AzureMachine status - https://github.com/niachary/cluster-api-provider-azure/blob/1b751e6b304186218f9b16f787a96230ddf796c4/cloud/scope/azurestackedgeippool.go#L162

Assuming the service crashes at this point and newAzureStackEdgeMachineService could not be called,
1) When the next reconcile cycle is triggred, AzureMachine.ReconciledIP will be true (since it has already got a free IP from the pool)
2) Hence it will not go ahead to get another Free IP
3) newAzureStackEdgeMachineService will be called again
4) This will create the network interface with this IP information.

So the IP will not be leaked even if the service crashes before calling newAzureStackEdgeMachineService.